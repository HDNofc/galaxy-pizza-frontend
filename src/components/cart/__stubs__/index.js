export const cartState = {
  items: [
    {
      cartId: 'cart-147',
      name: 'Мясное барбекю',
      description:
        'Мясная пицца с пикантной пепперони, альпийскими колбасками, Моцареллой, луком и соусом Барбекю',
      price: 1168,
      size: 35,
      dough: 'fat',
      crust: 'sausage',
      imageUrl: {
        crop: {
          jpg: 'https://api.papajohns.ru/images/catalog/thumbs/full/f945f964cd6a8f173ceaa66cc6cb3dbb.jpg',
          webp: 'https://api.papajohns.ru/images/catalog/thumbs/full/f945f964cd6a8f173ceaa66cc6cb3dbb.webp',
        },
        full: {
          jpg: 'https://api.papajohns.ru/images/catalog/thumbs/cart/64e568f97a17aa5caf0284511cde70a7.jpg',
          webp: 'https://api.papajohns.ru/images/catalog/thumbs/cart/64e568f97a17aa5caf0284511cde70a7.webp',
        },
      },
      amount: 2,
    },
    {
      cartId: 'cart-145',
      name: 'Маленькая италия',
      description:
        'Сыр моцарелла, пикантная пепперони, ароматная свинина, орегано, шампиньоны, черные оливки с фирменным томатным соусом',
      price: 639,
      size: 23,
      dough: 'slim',
      crust: 'none',
      imageUrl: {
        crop: {
          jpg: 'https://api.papajohns.ru/images/catalog/thumbs/full/Little-Italy-Thin.jpg',
          webp: 'https://api.papajohns.ru/images/catalog/thumbs/full/Little-Italy-Thin.webp',
        },
        full: {
          jpg: 'https://api.papajohns.ru/images/catalog/thumbs/cart/Little-Italy-Thin.jpg',
          webp: 'https://api.papajohns.ru/images/catalog/thumbs/cart/Little-Italy-Thin.webp',
        },
      },
      amount: 1,
    },
    {
      cartId: 'cart-170',
      name: 'Веганская с овощами',
      description:
        'Фирменный томатный соус, веганская Моцарелла, сладкий зелёный перец, шампиньоны, маслины, лук и сочные томаты',
      price: 799,
      size: 30,
      dough: 'fat',
      crust: 'none',
      imageUrl: {
        crop: {
          jpg: 'https://api.papajohns.ru/images/catalog/thumbs/full/9efdd0646d0a810e36e62854f71fe94b.jpg',
          webp: 'https://api.papajohns.ru/images/catalog/thumbs/full/9efdd0646d0a810e36e62854f71fe94b.webp',
        },
        full: {
          jpg: 'https://api.papajohns.ru/images/catalog/thumbs/cart/5f898b25ffeb3723fdc47fccc2a777a9.jpg',
          webp: 'https://api.papajohns.ru/images/catalog/thumbs/cart/5f898b25ffeb3723fdc47fccc2a777a9.webp',
        },
      },
      amount: 1,
    },
  ],
  totalCount: 4,
  totalPrice: 3774,
};
