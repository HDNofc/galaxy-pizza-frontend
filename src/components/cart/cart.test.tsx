import { within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import renderWithStore from 'test-utils/render-with-store';
import Cart from './cart';
import { cartState } from './__stubs__/';

const stateMock = { cart: cartState };

describe('<Cart>', () => {
  it('отображается корректно с верным количеством элементов', () => {
    const { container, getByText, getByRole } = renderWithStore(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
      { state: stateMock }
    );

    expect(container.firstChild).toMatchSnapshot();

    expect(container.querySelectorAll('li')).toHaveLength(3);
    expect(getByRole('heading', { hidden: true })).toHaveTextContent(/(4)/i);
    expect(getByText(/3 774 ₽/i)).toBeInTheDocument();
  });

  it('отображается заглушка, если корзина пуста', () => {
    const { getByText } = renderWithStore(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>
    );

    expect(getByText(/Корзина пуста/i)).toBeInTheDocument();
  });

  it('корзина очищается при клике на кнопку', () => {
    const { getByText, getByRole } = renderWithStore(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
      { state: stateMock }
    );

    expect(getByRole('heading', { hidden: true })).toHaveTextContent(/(4)/i);
    userEvent.click(getByText(/очистить/i));
    expect(getByText(/Корзина пуста/i)).toBeInTheDocument();
  });

  it('при увеличении/уменьшении количества заказываемых пицц изменяется сумма пицц в корзине', () => {
    const { container, getByRole } = renderWithStore(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
      { state: stateMock }
    );

    expect(container.querySelectorAll('li')).toHaveLength(3);
    expect(getByRole('heading', { hidden: true })).toHaveTextContent(/(4)/i);

    const firstPizza = container.querySelectorAll('li')[0];
    const incrementFirstPizzaButton = firstPizza.querySelectorAll('button')[2];
    const decrementFirstPizzaButton = firstPizza.querySelectorAll('button')[1];

    userEvent.click(incrementFirstPizzaButton);
    expect(getByRole('heading', { hidden: true })).toHaveTextContent(/(5)/i);

    userEvent.click(decrementFirstPizzaButton);
    userEvent.click(decrementFirstPizzaButton);
    expect(getByRole('heading', { hidden: true })).toHaveTextContent(/(3)/i);
  });

  it('при увеличении/уменьшении количества заказываемых пицц изменяется сумма оплаты', () => {
    const { container, getByText } = renderWithStore(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
      { state: stateMock }
    );

    expect(getByText(/3 774 ₽/i)).toBeInTheDocument();

    const firstPizza = container.querySelectorAll('li')[0];
    const incrementFirstPizzaButton = firstPizza.querySelectorAll('button')[2];
    const decrementFirstPizzaButton = firstPizza.querySelectorAll('button')[1];

    userEvent.click(incrementFirstPizzaButton);
    expect(getByText(/4 942 ₽/i)).toBeInTheDocument();

    userEvent.click(decrementFirstPizzaButton);
    userEvent.click(decrementFirstPizzaButton);
    expect(getByText(/2 606 ₽/i)).toBeInTheDocument();
  });

  it('увеличение/уменьшение количества конкретной пиццы происходит корректно', () => {
    const { container } = renderWithStore(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
      { state: stateMock }
    );

    expect(container.querySelectorAll('li')).toHaveLength(3);

    const firstPizza = container.querySelectorAll('li')[0];
    const incrementFirstPizzaButton = firstPizza.querySelectorAll('button')[2];
    const decrementFirstPizzaButton = firstPizza.querySelectorAll('button')[1];

    expect(within(firstPizza).getByText('2')).toBeInTheDocument();
    userEvent.click(incrementFirstPizzaButton);
    expect(within(firstPizza).queryByText('2')).not.toBeInTheDocument();
    expect(within(firstPizza).getByText('3')).toBeInTheDocument();

    userEvent.click(decrementFirstPizzaButton);
    userEvent.click(decrementFirstPizzaButton);
    expect(within(firstPizza).queryByText('3')).not.toBeInTheDocument();
    expect(within(firstPizza).getByText('1')).toBeInTheDocument();
  });

  it('при уменьшении количества конкретной пиццы ниже 1 происходит ее удаление', () => {
    const { container } = renderWithStore(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
      { state: stateMock }
    );

    expect(container.querySelectorAll('li')).toHaveLength(3);

    const firstPizza = container.querySelectorAll('li')[0];
    const decrementFirstPizzaButton = firstPizza.querySelectorAll('button')[1];

    expect(within(firstPizza).getByText('2')).toBeInTheDocument();
    userEvent.click(decrementFirstPizzaButton);
    userEvent.click(decrementFirstPizzaButton);
    expect(container.querySelectorAll('li')).toHaveLength(2);
  });

  it('удаление конкретной пиццы происходит корректно', () => {
    const { container } = renderWithStore(
      <MemoryRouter>
        <Cart />
      </MemoryRouter>,
      { state: stateMock }
    );

    expect(container.querySelectorAll('li')).toHaveLength(3);

    const firstPizza = container.querySelectorAll('li')[0];
    const removePizzaButton = within(firstPizza).getByText(/Удалить пиццу из корзины/i);

    userEvent.click(removePizzaButton);
    expect(container.querySelectorAll('li')).toHaveLength(2);
  });
});
