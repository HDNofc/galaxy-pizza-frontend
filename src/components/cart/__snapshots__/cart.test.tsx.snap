// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart> отображается корректно с верным количеством элементов 1`] = `
<div
  class="Cart-sc-1eqnpd4-0 gmhNp"
>
  <div
    class="Header-sc-1eqnpd4-1 iVioYV"
  >
    <h3
      class="Title-sc-1eqnpd4-2 cXAJfc"
    >
      Корзина 
      <span
        class="CartTotalCount-sc-1eqnpd4-3 eEFesY"
      >
        (
        4
        )
      </span>
    </h3>
    <button
      class="Button-sc-1fut0te-0 hlTeFB ClearCartButton-sc-1eqnpd4-4 fqAwVH"
      type="button"
    >
      <span
        class="Text-sc-1fut0te-1"
      >
        очистить
      </span>
    </button>
  </div>
  <ul
    class="CartList-sc-hb6owy-0 jBTWLT"
  >
    <li
      class="CartListItem-sc-hb6owy-1 bUxpHq"
    >
      <div
        class="ItemBody-sc-hb6owy-2 bCXgRP"
      >
        <div
          class="ImagePlaceholder-sc-hb6owy-4 dmvjGy"
        >
          <div
            class="ImagePlaceholder-sc-1lq6aod-0"
          >
            <div
              class="Placeholder-sc-1lq6aod-1 eizUTG"
            >
              <svg
                class="pizza-placeholder-image"
              >
                pizza-placeholder.svg
              </svg>
            </div>
            <div
              class="Content-sc-1lq6aod-2 iLQvsD"
            >
              <picture
                class="Picture-sc-hb6owy-3 lfkRa-D"
              >
                <source
                  srcset=""
                  type="image/webp"
                />
                <img
                  alt=""
                  class="CartListImage-sc-hb6owy-5 iHzSkv"
                  src=""
                />
              </picture>
            </div>
          </div>
        </div>
        <div
          class="ItemInfo-sc-hb6owy-6 gdDXea"
        >
          <span
            class="ItemName-sc-hb6owy-7 fZhgMN"
          >
            Мясное барбекю
          </span>
          <p
            class="ItemShortDescription-sc-hb6owy-8 eYepFn"
          >
            Традиционное тесто, 35 см, колбасный борт
          </p>
          <p
            class="ItemFullDescription-sc-hb6owy-9 bllVyv"
          >
            Мясная пицца с пикантной пепперони, альпийскими колбасками, Моцареллой, луком и соусом Барбекю
          </p>
        </div>
        <button
          class="Button-sc-1fut0te-0 jOZQxY DeleteButton-sc-hb6owy-10 elyAfU"
          type="button"
        >
          <span
            class="Text-sc-1fut0te-1"
          >
            Удалить пиццу из корзины
          </span>
          <svg
            class="cart-list-item-cross"
          >
            cross.svg
          </svg>
        </button>
      </div>
      <div
        class="ItemSummaryPrice-sc-hb6owy-11 czNIxm"
      >
        <div
          class="Counter-sc-9bdt3x-0 fPNYnq"
        >
          <button
            class="Button-sc-1fut0te-0 DIwKh"
            type="button"
          >
            <span
              class="Text-sc-1fut0te-1"
            >
              Отнять пиццу
            </span>
            <svg>
              minus.svg
            </svg>
          </button>
          <span
            class="Value-sc-9bdt3x-1 jOvzOn"
          >
            2
          </span>
          <button
            class="Button-sc-1fut0te-0 DIwKh"
            type="button"
          >
            <span
              class="Text-sc-1fut0te-1"
            >
              Добавить пиццу
            </span>
            <svg>
              plus.svg
            </svg>
          </button>
        </div>
        <span
          class="ItemPrice-sc-hb6owy-12 iQvgly"
        >
          1 168 ₽
        </span>
      </div>
    </li>
    <li
      class="CartListItem-sc-hb6owy-1 bUxpHq"
    >
      <div
        class="ItemBody-sc-hb6owy-2 bCXgRP"
      >
        <div
          class="ImagePlaceholder-sc-hb6owy-4 dmvjGy"
        >
          <div
            class="ImagePlaceholder-sc-1lq6aod-0"
          >
            <div
              class="Placeholder-sc-1lq6aod-1 eizUTG"
            >
              <svg
                class="pizza-placeholder-image"
              >
                pizza-placeholder.svg
              </svg>
            </div>
            <div
              class="Content-sc-1lq6aod-2 iLQvsD"
            >
              <picture
                class="Picture-sc-hb6owy-3 lfkRa-D"
              >
                <source
                  srcset=""
                  type="image/webp"
                />
                <img
                  alt=""
                  class="CartListImage-sc-hb6owy-5 iHzSkv"
                  src=""
                />
              </picture>
            </div>
          </div>
        </div>
        <div
          class="ItemInfo-sc-hb6owy-6 gdDXea"
        >
          <span
            class="ItemName-sc-hb6owy-7 fZhgMN"
          >
            Маленькая италия
          </span>
          <p
            class="ItemShortDescription-sc-hb6owy-8 eYepFn"
          >
            Тонкое тесто, 23 см
          </p>
          <p
            class="ItemFullDescription-sc-hb6owy-9 bllVyv"
          >
            Сыр моцарелла, пикантная пепперони, ароматная свинина, орегано, шампиньоны, черные оливки с фирменным томатным соусом
          </p>
        </div>
        <button
          class="Button-sc-1fut0te-0 jOZQxY DeleteButton-sc-hb6owy-10 elyAfU"
          type="button"
        >
          <span
            class="Text-sc-1fut0te-1"
          >
            Удалить пиццу из корзины
          </span>
          <svg
            class="cart-list-item-cross"
          >
            cross.svg
          </svg>
        </button>
      </div>
      <div
        class="ItemSummaryPrice-sc-hb6owy-11 czNIxm"
      >
        <div
          class="Counter-sc-9bdt3x-0 fPNYnq"
        >
          <button
            class="Button-sc-1fut0te-0 DIwKh"
            type="button"
          >
            <span
              class="Text-sc-1fut0te-1"
            >
              Отнять пиццу
            </span>
            <svg>
              minus.svg
            </svg>
          </button>
          <span
            class="Value-sc-9bdt3x-1 jOvzOn"
          >
            1
          </span>
          <button
            class="Button-sc-1fut0te-0 DIwKh"
            type="button"
          >
            <span
              class="Text-sc-1fut0te-1"
            >
              Добавить пиццу
            </span>
            <svg>
              plus.svg
            </svg>
          </button>
        </div>
        <span
          class="ItemPrice-sc-hb6owy-12 iQvgly"
        >
          639 ₽
        </span>
      </div>
    </li>
    <li
      class="CartListItem-sc-hb6owy-1 bUxpHq"
    >
      <div
        class="ItemBody-sc-hb6owy-2 bCXgRP"
      >
        <div
          class="ImagePlaceholder-sc-hb6owy-4 dmvjGy"
        >
          <div
            class="ImagePlaceholder-sc-1lq6aod-0"
          >
            <div
              class="Placeholder-sc-1lq6aod-1 eizUTG"
            >
              <svg
                class="pizza-placeholder-image"
              >
                pizza-placeholder.svg
              </svg>
            </div>
            <div
              class="Content-sc-1lq6aod-2 iLQvsD"
            >
              <picture
                class="Picture-sc-hb6owy-3 lfkRa-D"
              >
                <source
                  srcset=""
                  type="image/webp"
                />
                <img
                  alt=""
                  class="CartListImage-sc-hb6owy-5 iHzSkv"
                  src=""
                />
              </picture>
            </div>
          </div>
        </div>
        <div
          class="ItemInfo-sc-hb6owy-6 gdDXea"
        >
          <span
            class="ItemName-sc-hb6owy-7 fZhgMN"
          >
            Веганская с овощами
          </span>
          <p
            class="ItemShortDescription-sc-hb6owy-8 eYepFn"
          >
            Традиционное тесто, 30 см
          </p>
          <p
            class="ItemFullDescription-sc-hb6owy-9 bllVyv"
          >
            Фирменный томатный соус, веганская Моцарелла, сладкий зелёный перец, шампиньоны, маслины, лук и сочные томаты
          </p>
        </div>
        <button
          class="Button-sc-1fut0te-0 jOZQxY DeleteButton-sc-hb6owy-10 elyAfU"
          type="button"
        >
          <span
            class="Text-sc-1fut0te-1"
          >
            Удалить пиццу из корзины
          </span>
          <svg
            class="cart-list-item-cross"
          >
            cross.svg
          </svg>
        </button>
      </div>
      <div
        class="ItemSummaryPrice-sc-hb6owy-11 czNIxm"
      >
        <div
          class="Counter-sc-9bdt3x-0 fPNYnq"
        >
          <button
            class="Button-sc-1fut0te-0 DIwKh"
            type="button"
          >
            <span
              class="Text-sc-1fut0te-1"
            >
              Отнять пиццу
            </span>
            <svg>
              minus.svg
            </svg>
          </button>
          <span
            class="Value-sc-9bdt3x-1 jOvzOn"
          >
            1
          </span>
          <button
            class="Button-sc-1fut0te-0 DIwKh"
            type="button"
          >
            <span
              class="Text-sc-1fut0te-1"
            >
              Добавить пиццу
            </span>
            <svg>
              plus.svg
            </svg>
          </button>
        </div>
        <span
          class="ItemPrice-sc-hb6owy-12 iQvgly"
        >
          799 ₽
        </span>
      </div>
    </li>
  </ul>
  <div
    class="Footer-sc-1eqnpd4-6 dutOFD"
  >
    <div
      class="PriceWrapper-sc-1eqnpd4-7 gTiDIp"
    >
      <span
        class="TotalPriceText-sc-1eqnpd4-8 bStYHF"
      >
        Сумма заказа:
      </span>
       
      <span
        class="TotalPrice-sc-1eqnpd4-9 hABHep"
      >
        3 774 ₽
      </span>
    </div>
    <a
      class="Link-sc-16ffk5r-0 kUJqkA Button-sc-1fut0te-0 OrderButton-sc-1eqnpd4-11 kDjXIb kpLaUN"
      href="/order"
    >
      Оформить
    </a>
  </div>
</div>
`;
